staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/admission"
+	admissionapi "k8s.io/pod-security-admission/admission/api"
+	"k8s.io/pod-security-admission/admission/api/validation"
+	"k8s.io/pod-security-admission/api"
+	"k8s.io/pod-security-admission/metrics"
+	"k8s.io/pod-security-admission/policy"
+)
+
+const (
+	namespaceMaxPodsToCheck  = 3000
+	namespacePodCheckTimeout = 1 * time.Second
+)
+
+// Admission implements the core admission logic for the Pod Security Admission controller.
+// The admission logic can be
+type Admission struct {
+	Configuration *admissionapi.PodSecurityConfiguration
+
+	// Getting policy checks per level/version
+	Registry policy.Registry
+
+	// Metrics
+	Metrics metrics.EvaluationRecorder
+
+	// Arbitrary object --> PodSpec
+	PodSpecExtractor PodSpecExtractor
+
+	// API connections
+	NamespaceGetter NamespaceGetter
```

Comment:
note to self, come back to ensure this punches through on misses, possibly using the LRU cache approach we used elsewhere

Reply:
the implementation we're using does:

https://github.com/kubernetes/kubernetes/blob/eaa0a432f56e732473dc79fdcb370d50de70577d/staging/src/k8s.io/pod-security-admission/admission/namespace.go#L42-L50